docs/design.md

```diff
@@ -0,0 +1,273 @@
+  - [Comparing with Ansible Operator](#comparing-with-ansible-operator)
+  - [More Considerations](#more-considerations)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+# The Design of Crossplane Provider for Ansible
+
+## Overview
+
+Crossplane is often used to assemble infrastructure from multiple vendors declaratively in Kubernetes which is a typical cloud native architecture. However, this may not be true for many real-world organizations who may have many IT systems that are traditional based and have invested heavily on management automation for these systems.
+
+Ansible, as a popular automation technology, has a large user base and mature eco-system. It is widely adopted by many organizations to automate management for different varieties of IT systems, ranging from cloud native to non cloud native.
+
+The Crossplane provider for Ansible is targeted to extend the Crossplane scope by enabling its integration with Ansible to build a bridge between the cloud native and non cloud native worlds, open the door to drive and reuse existing automation assets, no matter it is cloud native or non cloud native, to manage hybrid technologies using the same way consistently and help organizations transition to cloud native while keep their existing investments.
+
+## How It Works
+
+The high level architecture for Ansible provider is illustrated as below.
+
+![](images/ansible-provider-design.png)
+
+Ansible provider uses Kubernetes resource `AnsibleRun` to represent a certain Ansible run, e.g.: to run a set of Ansible roles or playbooks. The `AnsibleRun` resource also refers to a `ProviderConfig` resource including all information at provider level that are needed to support Ansible runs, e.g.: credentials needed to access a remote place that hosts Ansible contents, requirements needed to install as prerequisite before run Ansible contents. Regarding the Ansible contents, it can be collections, roles, or playbooks that are hosted in Ansible Galaxy, public or private Automation Hub, or GitHub repository.
+
+As the managed resource of Ansible provider, once the `AnsibleRun` resource is defined, the provider will download the Ansible contents into the provider working directory first via the method `Connect()` in AnsibleRun controller, then trigger the Ansible run according to the provider lifecycle via method `Observe()/Create()/Update()/Delete()`. 
+
+Inside the provider, it actually relies on the command line tool [`ansible-galaxy`](https://docs.ansible.com/ansible/latest/cli/ansible-galaxy.html) to download Ansible contents, then use [`ansible-runner`](https://ansible-runner.readthedocs.io/) to execute them. The module [`ansible-runner-http`](https://github.com/ansible/ansible-runner-http) is a plugin for ansible runner that allows to emit Ansible status and events to HTTP services in the form of POST events, so as to notify AnsibleRun controller about the run results and attatch that to `AnsibleRun` as status.
+
```

Comment:
do we really need a plugin to get `ansible-runner` execution status. the ansible-runner output and exit code(s) are not enough to decide the state of execution ?

Reply:
It is just referenced from ansible operator where it use the `ansible-runner-http` plugin to collect the results. We can leave as is for now and add later when there is a need, suppose the design doc is written for the final picture :-)